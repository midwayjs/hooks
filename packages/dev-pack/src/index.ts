export * from './middleware'
export * from './server'
export * from './share'
