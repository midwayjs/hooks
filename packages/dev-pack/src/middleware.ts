import { NextFunction, Request, Response } from 'express'
import { CreateOptions, DevServer } from './server'
import { createProxyMiddleware } from 'http-proxy-middleware'
import { parse } from 'url'
import { ServerState } from './share'
import { createDebug } from '@midwayjs/hooks-core'

const debug = createDebug('hooks-dev-pack: middleware')

export async function createExpressDevPack(options: CreateOptions) {
  options.cwd ??= process.cwd()
  options.sourceDir ??= 'src/api'

  const server = new DevServer(options)
  debug('options', options)
  await server.run()

  const proxy = createProxyMiddleware({
    target: `http://localhost:${server.port}`,
    logLevel: 'error',
    followRedirects: true,
  })

  const MidwayHooksDevPackMiddleware = async (
    req: Request,
    res: Response,
    next: NextFunction
  ) => {
    const path = parse(req.url).pathname

    const internalServerError = () => {
      res.statusCode = 500
      const send = typeof res.send === 'function' ? res.send : res.end
      send.call(
        res,
        `Server error, please check the console output and fix the problem\n服务器错误，请查看控制台输出并修复问题`
      )
    }

    if (server.state === ServerState.Error) {
      return internalServerError()
    }

    try {
      await server.ready()
    } catch {
      return internalServerError()
    }

    if (await server.isMatch(path, req.method)) {
      proxy(req, res, next)
    } else {
      next()
    }
  }

  return {
    middleware: MidwayHooksDevPackMiddleware,
    server,
  }
}
