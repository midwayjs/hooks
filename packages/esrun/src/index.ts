import { readDefaultTsConfig } from '@swc-node/register/read-default-tsconfig'
import { register as registerSWC } from '@swc-node/register/register'

export function register() {
  const options = readDefaultTsConfig()

  // ensure all files are compiled
  options.files = null
  registerSWC(options)

  if (options.paths && Object.keys(options.paths).length) {
    require('tsconfig-paths/register')
  }
}
