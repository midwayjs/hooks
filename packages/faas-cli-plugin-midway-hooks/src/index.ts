export { MidwayHooksPlugin } from './plugin'
export { ViewRoutesPlugin } from './routes'
// export { InstallPlugin } from './installer'
