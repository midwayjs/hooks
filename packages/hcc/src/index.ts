import { getRouter, getSource, setProjectRoot } from '@midwayjs/hooks-internal'
import { GenerateTarget, getEntryCode } from './midway'

export async function getPreloadCode(cwd: string) {
  setProjectRoot(cwd)

  const source = getSource({ useSourceFile: false })
  const router = getRouter(source)

  return `
    ${getEntryCode(source, router, GenerateTarget.JS)}
    module.exports = require('./configuration');
  `
}
