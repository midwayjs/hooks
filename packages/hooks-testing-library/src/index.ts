import { join } from 'path'

import { IMidwayApplication, IMidwayFramework } from '@midwayjs/core'
import {
  ApiFunction,
  getConfig,
  getProjectRoot,
  ProjectConfig,
  validateFunction,
} from '@midwayjs/hooks-core'
import {
  close,
  createApp as createWebApp,
  createFunctionApp as createFaaSApp,
  createHttpRequest,
} from '@midwayjs/mock'

export type CreateAppOption = {
  root?: string
}

async function createAppImplementation(
  options: CreateAppOption,
  isFunction: boolean
) {
  const root = getProjectRoot(options?.root ?? (global as any).testPath)
  const config = getConfig(root)

  if (options?.root) {
    delete options?.root
  }

  const cwd = process.cwd()
  process.chdir(root)

  const creator = isFunction ? createFaaSApp : createWebApp

  const app: IMidwayApplication<any> = await creator(
    root,
    Object.assign(
      {
        baseDir: join(root, config.source),
      },
      options
    )
  )

  process.chdir(cwd)
  return new HooksApplication(app, config)
}

export async function createApp<
  T extends IMidwayFramework<any, U>,
  U = T['configurationOptions'] & CreateAppOption
>(options?: U) {
  return createAppImplementation(options, false)
}

export async function createFunctionApp<
  T extends IMidwayFramework<any, U>,
  U = T['configurationOptions'] & CreateAppOption
>(options?: U) {
  return createAppImplementation(options, true)
}

export class HooksApplication {
  private readonly app: IMidwayApplication<any>
  private config: ProjectConfig
  constructor(app: IMidwayApplication<any>, config: ProjectConfig) {
    this.app = app
    this.config = config
  }

  async runFunction<T extends ApiFunction>(
    fn: T,
    ...args: Parameters<T>
  ): Promise<ReturnType<T>> {
    validateFunction(fn, 'fn')

    const response = await this.request(fn, ...args)

    if (!response.ok) {
      throw response.body
    }

    if (response.type === 'application/json') {
      return response.body
    }

    return response.text
  }

  request<T extends ApiFunction>(fn: T, ...args: Parameters<T>) {
    validateFunction(fn, 'fn')

    const supertest = createHttpRequest(this.app)
    if (args.length === 0) {
      return supertest.get(fn._param.url)
    }

    return supertest
      .post(fn._param.url)
      .send({ args })
      .set('Content-Type', 'application/json')
  }

  async close() {
    await close(this.app)
  }
}
