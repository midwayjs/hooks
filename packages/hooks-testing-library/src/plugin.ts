import { SuperAgentRequest } from 'superagent'

import { superjson } from '@midwayjs/hooks-core'

export function SuperJSONPlugin(enableSuperjson: boolean) {
  if (!enableSuperjson) {
    return () => {}
  }
  return (req: SuperAgentRequest) => {
    req.on('response', (res) => {
      if (res.body && res.type.includes('application/json')) {
        res.body = superjson.deserialize(res.body)
      }
    })
  }
}
