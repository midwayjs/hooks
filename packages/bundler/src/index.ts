import { createUnplugin, UnpluginOptions } from 'unplugin'
import {
  AbstractRouter,
  ApiRoute,
  createDebug,
  EXPORT_DEFAULT_FUNCTION_ALIAS,
  parseApiModule,
} from '@midwayjs/hooks-core'
import '@midwayjs/esrun/register'

export { EsbuildPlugin, VitePlugin, RollupPlugin } from 'unplugin'

const debug = createDebug('hooks-bundler')

const omit = (key: string, { [key]: _, ...obj }) => obj

interface BundlerConfig {
  name: string
  router: AbstractRouter
}

export abstract class AbstractBundlerAdapter {
  constructor(private config: BundlerConfig) {}

  getName() {
    return this.config.name
  }

  abstract getSource(): string

  getRouter() {
    return this.config.router
  }

  getUnplugin(): Partial<UnpluginOptions> {
    return {}
  }

  transformApiRoutes(apis: ApiRoute[]) {
    return apis
  }

  generateClient(apis: ApiRoute[]) {
    let fetcherId = 0
    const importClient = new Map<string, [string, string, number]>()

    const importCodes = new Set<string>()
    const functionCodes = []

    for (const {
      trigger,
      functionName,
      functionId,
      useInputMetadata,
    } of apis) {
      if (!trigger.requestClient) continue
      const { fetcher, client } = trigger.requestClient

      const clientId = `${client}_${fetcher}`
      if (!importClient.has(clientId)) {
        importClient.set(clientId, [client, fetcher, fetcherId++])
      }

      const [clientName, fetcherName, idx] = importClient.get(clientId)
      const uniqueFetcher = `rpc$${idx}`
      importCodes.add(
        `import { ${fetcherName} as ${uniqueFetcher} } from '${clientName}'`
      )

      const exportModifier =
        functionName === EXPORT_DEFAULT_FUNCTION_ALIAS ? 'default' : ''
      functionCodes.push(`
        export ${exportModifier} function ${functionName} (...args) {
          const route = ${functionName}.route;
          return ${uniqueFetcher}(...args, route)
        };
        ${functionName}.route = {
          trigger: ${JSON.stringify(omit('requestClient', trigger))},
          functionId: '${functionId}',
          useInputMetadata: ${!!useInputMetadata},
        };
      `)
    }

    const code = [
      '// This file is auto-generated by @midwayjs/bundler. Any changes will be overwritten.',
      ...importCodes,
      ...functionCodes,
    ].join('\n')

    debug('generate client code: %s', code)

    return code
  }
}

export function createBundlerPlugin(adapter: AbstractBundlerAdapter) {
  return createUnplugin((options) => {
    return {
      name: adapter.getName(),
      enforce: 'pre',
      transformInclude(id: string) {
        const source = adapter.getSource()
        const router = adapter.getRouter()

        if (!router.isSourceFile(id, source)) {
          return false
        }

        return router.isApiFile({
          file: id,
          mod: requireWithoutCache(id),
        })
      },
      async transform(code: string, id: string) {
        const apis = parseApiModule(
          requireWithoutCache(id),
          id,
          adapter.getRouter()
        )
        const transformedApis = adapter.transformApiRoutes(apis)
        return adapter.generateClient(transformedApis)
      },
      ...adapter.getUnplugin(),
    }
  })
}

/**
 * In dev mode, ensure get latest module content when file changes
 */
export function requireWithoutCache(module: string) {
  delete require.cache[require.resolve(module)]
  return require(module)
}
