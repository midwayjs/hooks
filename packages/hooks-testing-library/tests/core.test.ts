import path from 'path'

import { createApp, HooksApplication } from '../src'
import {
  returnString,
  returnObject,
  getContext,
  get,
  post,
} from './fixtures/core/src/lambda/index'
import { tryInject, tryInjectLocal } from './fixtures/core/src/lambda/ioc'
import middleware from './fixtures/core/src/lambda/middleware'
import { underscore } from './fixtures/core/src/underscore'

let app: HooksApplication
beforeAll(async () => {
  app = await createApp({
    root: path.join(__dirname, './fixtures/core'),
  })
})

afterAll(async () => {
  await app.close()
})

describe('Hooks Api', () => {
  test('HTTP GET', async () => {
    expect(await app.runFunction(get)).toMatchSnapshot()
  })

  test('HTTP POST', async () => {
    expect(await app.runFunction(post, 'Lxxyx')).toMatchSnapshot()
  })

  test('return string', async () => {
    expect(await app.runFunction(returnString)).toMatchSnapshot()
  })

  test('return object', async () => {
    expect(await app.runFunction(returnObject)).toMatchSnapshot()
  })

  test('get request context', async () => {
    expect(await app.runFunction(getContext)).toMatchSnapshot()
  })

  // TODO Handle HTTP Error

  test('underscore', async () => {
    const { path } = await app.runFunction(underscore)
    return expect(path).toEqual('/underscore/_underscore')
  })

  test('middleware', async () => {
    const response = await app.request(middleware).expect(200)
    // Set by middleware
    expect(response.headers.global).toBeTruthy()
    expect(response.headers.file).toBeTruthy()
    expect(response.headers.api).toBeTruthy()
  })

  test('IoC', async () => {
    expect(await app.runFunction(tryInject)).toMatchSnapshot()
    expect(await app.runFunction(tryInjectLocal)).toMatchSnapshot()
  })
})
