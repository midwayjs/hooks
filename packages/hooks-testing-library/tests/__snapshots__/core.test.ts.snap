// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hooks Api HTTP GET 1`] = `"GET"`;

exports[`Hooks Api HTTP POST 1`] = `
Object {
  "method": "POST",
  "name": "Lxxyx",
}
`;

exports[`Hooks Api IoC 1`] = `
Array [
  "Miao~",
  "Miao~",
  "Miao~",
]
`;

exports[`Hooks Api IoC 2`] = `
Object {
  "find": Array [
    "Local",
  ],
  "findAll": Array [
    "Miao~",
    "Miao~",
    "Miao~",
  ],
}
`;

exports[`Hooks Api get request context 1`] = `
Object {
  "path": "/api/getContext",
}
`;

exports[`Hooks Api return object 1`] = `
Object {
  "message": "Hello World!",
}
`;

exports[`Hooks Api return string 1`] = `"Hello World"`;
